url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/GeneticAlgorithm.java

```diff
@@ -0,0 +1,504 @@
+		// a probability of 2:3 and 1:3 respectively, which genes to pass on.
+		else if(mother.levelOfTonality == 2 && father.levelOfTonality == 1){
+			if(randTwoDom == 0 || randTwoDom == 1)
+				child = new Chromosome(this.minOctave, this.maxOctave, mother.getKeyGene(), mother.getScaleGene());
+			else
+				child = new Chromosome(this.minOctave, this.maxOctave, father.getKeyGene(), mother.getScaleGene());
+		}
+		else if(mother.levelOfTonality == 1 && father.levelOfTonality == 2){
+			if(randTwoDom == 0 || randTwoDom == 1)
+				child = new Chromosome(this.minOctave, this.maxOctave, father.getKeyGene(), father.getScaleGene());
+			else
+				child = new Chromosome(this.minOctave, this.maxOctave, mother.getKeyGene(), father.getScaleGene());
+			
+		}
+		
+		// If Both parents have two dominant tonal genes randomly select which of the two to pass on
+		else if(mother.levelOfTonality == 2 && father.levelOfTonality == 2){
+			switch(rand){
+			
+			case 1:
+				child = new Chromosome(this.minOctave, this.maxOctave, mother.getKeyGene(), father.getScaleGene());
+				break;
+			case 2:
+				child = new Chromosome(this.minOctave, this.maxOctave, father.getKeyGene(), mother.getScaleGene());
+				break;
+			}
+			
+		}
+		// Crossover for both types of Dynamic Gene
+		switch(rand){
+		case 1:
+			child.setDynamicGene(mother.getDynamicGene());
+			child.setDshapeGene(father.getDshapeGene());
+			break;
+		case 2:
+			child.setDynamicGene(father.getDynamicGene());
+			child.setDshapeGene(mother.getDshapeGene());
+			break;
+		}
+		
+		// Crossover of Rhythmic Genes..
+		
+		// If crossover occurs before sorting the arrays the rhythmic blocks can 
+		// overflow a bar length ie if the second block of one parent is a 2 beat 
+		// block then if crossover occurs there it can end up leaving the child with
+		// 5 or 6 beats. So to avoid this, both parents are sorted and the rand variable
+		// selects one to be reversed so that 2 beat blocks that the child chromosome
+		// inherits aren't always the end block.
+		switch(rand){
+		case 1:
+			mother.sortRhythm();
+			Collections.reverse(mother.rhythmGenes);
+			father.sortRhythm();
+			break;
+		case 2:
+			mother.sortRhythm();
+			father.sortRhythm();
+			Collections.reverse(father.rhythmGenes);
+			break;
+		}
+		// Store the parents rhythmic genes in two ArrayLists
+		ArrayList <Rhythmic_Block> mothersRhythm = mother.rhythmGenes;
+		ArrayList <Rhythmic_Block> fathersRhythm = father.rhythmGenes;
+		
+		ArrayList <Rhythmic_Block> childsRhythm = new ArrayList <Rhythmic_Block>();
+		
+		int sum = 0;
+		
+		// The arraylists are then crossed over, the crossover point is always at the 
+		// start of the third beat of bar in the set of rhythmic_blocks.
+		// The rand variable decides which parent has their first 2 beats crossed over.
+		switch(rand){
+		case 1:
+			for(Rhythmic_Block rb : mothersRhythm){
+				sum += rb.getBeatValue();
+				if(sum <= 2){
+					childsRhythm.add(rb);
+				}
+				else
+					break;
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 299:306, 317:324.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.